community/index/src/test/java/org/neo4j/index/gbptree/SeekCursorTest.java

```diff
@@ -990,6 +995,144 @@ public void shouldFailChildPointerIfReadFailureNotCausedByCheckpoint() throws Ex
         }
     }
 
+    @Test
+    public void shouldCatchupRootWhenRootNodeHasTooNewGeneration() throws Exception
+    {
+        // given
+        long id = cursor.getCurrentPageId();
+        long gen = node.gen( cursor );
+        MutableBoolean triggered = new MutableBoolean( false );
+        Supplier<Root> rootCatchup = () ->
+        {
+            triggered.setTrue();
+            return new Root( id, gen );
+        };
+
+        // when
+        try ( SeekCursor<MutableLong,MutableLong> seek = new SeekCursor<>( cursor, readKey, readValue, node, from, to,
+                layout, stableGen, unstableGen, generationSupplier, rootCatchup, gen - 1 ) )
+        {
+            // do nothing
+        }
+
+        // then
+        assertTrue( triggered.getValue() );
+    }
+
+    @Test
+    public void shouldCatchupRootWhenNodeHasTooNewGenerationWhileTraversingDownTree() throws Exception
+    {
+        // given
+        long gen = node.gen( cursor );
+        MutableBoolean triggered = new MutableBoolean( false );
+
+        // a newer leaf
+        long leftChild = cursor.getCurrentPageId();
+        node.initializeLeaf( cursor, stableGen + 1, unstableGen + 1 ); // A newer leaf
+        cursor.next();
+
+        // a root
+        long rootId = cursor.getCurrentPageId();
+        node.initializeInternal( cursor, stableGen, unstableGen );
+        long keyInRoot = 10L;
+        insertKey.setValue( keyInRoot );
+        node.insertKeyAt( cursor, insertKey, 0, 0, tmp );
+        node.setKeyCount( cursor, 1 );
+        // with old pointer to child (simulating reuse of child node)
+        node.setChildAt( cursor, leftChild, 0, stableGen, unstableGen );
+
+        // a root catchup that records usage
+        Supplier<Root> rootCatchup = () ->
+        {
+            try
+            {
+                triggered.setTrue();
+
+                // and set child generation to match pointer
+                cursor.next( leftChild );
+                cursor.zapPage();
+                node.initializeLeaf( cursor, stableGen, unstableGen );
+
+                cursor.next( rootId );
+                return new Root( rootId, gen );
+            }
+            catch ( IOException e )
+            {
+                throw new RuntimeException( e );
```

Comment:
its not that important here, but in general for future cases like that its better to use UncheckedIOException